### ISSUE TITLE
I want to use YARN on spark, whats quickest way to do it with flintrock, do you suggest some AMI ?

### ISSUE THREAD

**Issue by raghav20** [nchammas/flintrock#175]
If you are reporting a potential bug, include as many of the following
as are relevant (and delete this sentence, of course):
* Flintrock version: Latest
* Python version: 3
* OS: Linux

**Comment by nchammas** [comment]
Flintrock only supports running Spark in standalone mode, and there are no plans at the moment to support running on YARN. I don't think simply selecting a different AMI will help with Flintrock, since Flintrock needs to be aware of YARN in order to configure things correctly.

If you want to run Spark on YARN, I recommend looking at [spark-ec2](https://github.com/amplab/spark-ec2), which has an option for that, or looking at [EMR](https://aws.amazon.com/emr/), which runs Spark on YARN by default.

**Comment by xiandong79** [comment]
What about now?

**Comment by amritanshujain** [comment]
Including the feature of spark on YARN in flintrock would be great, I tried my hands on including the feature but hit some roadblocks many times. I am willing to contribute to this feature, if possible can you help me get started?

**Comment by amritanshujain** [comment]
Hi @nchammas 
Me and @ramchandra18 were working on this together. I have some commits on my fork too, we tried to setup the YARN environment and the configuration files and launching the cluster I guess we have been successful in running Spark on YARN using flintrock. I will push all the changes on my fork and if possible you can try and use it to see if its acceptable. We will further work on the tests :smile:

**Comment by ramchandra18** [comment]
@nchammas I would like you to review the code changes i have made in my fork. I just tweaked the existing class member functions replacing the standalone shell script calls with yarn daemon script call. Would be changing that and add a new function that is specific to yarn. Me and @amritanshujain would be working further to add the test cases. I am willing to work through the back and forth of the review process.

**Comment by nchammas** [comment]
OK, sounds good. Whenever y'all are ready, please open a PR and I will review your work. :+1:

**Comment by ramchandra18** [comment]
Hi @nchammas. Is the sftp used to copy the templates and download scripts done in parallel ?. Isn't there any GIL lock which prevents the threads to read them in parallel ?. I think that ThreadPoolExecutor wouldn't help with the GIL lock right ?

**Comment by nchammas** [comment]
The GIL doesn't affect IO-bound work done on threads the way it does with CPU-bound work, so that shouldn't be an issue here. Also, I don't see how this is related to adding YARN support.

**Comment by xiandong79** [comment]
Hi, @ramchandra18. Is **NodeLabel feature** of "Capacity Scheduler in YARN" in your branch available currently?

**Comment by ramchandra18** [comment]
@nchammas I have added few templates and found that the IO work is becoming a bottleneck.

**Comment by ramchandra18** [comment]
@xiandong79 I haven't yet added the NodeLabel feature. I will be doing it soon.

**Comment by nchammas** [comment]
@ramchandra18 - Interesting. What leads you to believe that? Please provide specifics.

Flintrock has been used to launch clusters with several hundred nodes, and sending up the config files has never been an issue. Unless YARN adds several MB worth of config files, I doubt that sending up the config files is an issue.

If you're seeing any launch slowness, it's more likely that that is caused by the slow Apache mirrors hosting Spark and Hadoop. That should be clear from the launch output. If so, the solution is to specify  a custom download source that is fast for your region.

**Comment by ramchandra18** [comment]
@nchammas It's my mistake. It wasn't the copying part that is consuming time. The download time of hadoop and spark packages is different for nodes. One of the node is consuming 2x time than the other nodes. I am using the default download source.

**Comment by ramchandra18** [comment]
@nchammas As suggested by you I would use my s3 bucket as a download source :smiley:

**Comment by xiandong79** [comment]
@ramchandra18 Hi, is your repo available to test?  I want to submit jobs using YARN scheduler.

**Comment by skyer9** [comment]
I did some work for enabling yarn cluster mode.

**To use default value of yarn configuration, the memory size of EC2 instance should be 16G or more.**

```sh
vi yarn-site.xml
```

```xml
<configuration>

<property>
    <name>yarn.resourcemanager.hostname</name>
    <value><private ip of master node></value>
</property>

</configuration>
```

```sh
vi enable-yarn.sh
```

```sh
#!/bin/sh

export HADOOP_PREFIX=/home/ec2-user/hadoop

echo "export HADOOP_PREFIX=$HADOOP_PREFIX" >> ~/.bashrc
echo "export HADOOP_HOME=$HADOOP_PREFIX" >> ~/.bashrc
echo "export HADOOP_COMMON_HOME=$HADOOP_PREFIX" >> ~/.bashrc
echo "export HADOOP_CONF_DIR=$HADOOP_PREFIX/etc/hadoop" >> ~/.bashrc
echo "export HADOOP_HDFS_HOME=$HADOOP_PREFIX" >> ~/.bashrc
echo "export HADOOP_MAPRED_HOME=$HADOOP_PREFIX" >> ~/.bashrc
echo "export HADOOP_YARN_HOME=$HADOOP_PREFIX" >> ~/.bashrc
```

copy file to cluster.

```sh
flintrock copy-file bigdata-cluster yarn-site.xml ./hadoop/etc/hadoop/
flintrock copy-file bigdata-cluster enable-yarn.sh ./
flintrock run-command bigdata-cluster 'sh ~/enable-yarn.sh'
```

ONLY IN MASTER NODE.

```sh
flintrock login bigdata-cluster
$HADOOP_PREFIX/sbin/yarn-daemon.sh start resourcemanager
exit
```

all node.

```sh
flintrock run-command bigdata-cluster '$HADOOP_PREFIX/sbin/yarn-daemon.sh start nodemanager'
```

check working.

```sh
vi hello.py
```

```python
# -*- coding: utf-8 -*-
from pyspark.sql import SparkSession
spark = SparkSession.builder.appName('abc').getOrCreate()

print('Hello, World!')
```

```sh
flintrock copy-file bigdata-cluster hello.py ./
flintrock login bigdata-cluster
spark-submit --master yarn \
             --deploy-mode cluster \
               hello.py
```
